build: correct ELF info for libgeos

otan opened the pull request:
Looks like cross compiling with rpath does not work, so install patchelf
and use them on the given .so files such that dlopen works using the
regular pattern.

Refs #54841

Release note (general change, bug fix): Fixed the rpath and so names of
libgeos.so and libgeos_c.so such that a dlopen to libgeos.so is not
needed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55129)
<!-- Reviewable:end -->

otan committed:
build: correct ELF info for libgeos

Looks like cross compiling with rpath does not work, so install patchelf
and use them on the given .so files such that dlopen works using the
regular pattern.

Release note (general change, bug fix): Fixed the rpath and so names of
libgeos.so and libgeos_c.so such that a dlopen to libgeos.so is not
needed.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I'm going to have to trust that you did the appropriate testing of this change, because I'm science dogging it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55129#-:-MI_bcr533J0yuoyBvFW:b-y1h9on)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)



<!-- Sent from Reviewable.io -->

otan commented:
> because I'm science dogging it.

😂

otan commented:
bors r=petermattis

timgraham commented:
Sorry, I didn't follow up and test this. With the libgeos files in CockroachDB 20.2.1, I get this error, whether or not the symlink is present.
```
Traceback (most recent call last):
...
  File "/home/tim/code/django/django/utils/functional.py", line 376, in _setup
    self._wrapped = self._setupfunc()
  File "/home/tim/code/django/django/contrib/gis/geos/libgeos.py", line 61, in load_geos
    _lgeos = CDLL(lib_path)
  File "/opt/python3.9.0/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libgeos.so.3.8.1: cannot open shared object file: No such file or directory
````

With the symlink, the setup looks like this:
```
$ ls -l /usr/local/lib/cockroach
total 428
-rwxr-xr-x 1 tim  tim  435520 Nov 24 18:20 libgeos_c.so
-rwxr-xr-x 1 tim  tim       0 Nov 24 18:23 libgeos.so
lrwxrwxrwx 1 root root     35 Nov 24 18:22 libgeos.so.3.8.1 -> /usr/local/lib/cockroach/libgeos.so
````

otan commented:
this will be available in 21.1, i could backport it but not sure how urgent it is as i wasn't really banking on this usage of the libgeos we ship....

if the test harness can run patchelf in the interim that'd be great!

timgraham commented:
Not a problem. I'll recommend using the distro provided version of libgeos for now.

timgraham commented:
I've confirmed this works with Django in Cockroach 21.1 beta 1. Thanks!
